src/bento/components/bento-autocomplete/1.0/component.tsx

```diff
@@ -0,0 +1,555 @@
+      if (
+        highlightUserEntry &&
+        substring.length &&
+        substring.length <= item.length &&
+        includes(lowerCaseItem, lowerCaseSubstring)
+      ) {
+        const substringStart = lowerCaseItem.indexOf(lowerCaseSubstring);
+        const substringEnd = substringStart + lowerCaseSubstring.length;
+        return (
+          <>
+            {item.slice(0, substringStart)}
+            <span class="autocomplete-partial">
+              {item.slice(substringStart, substringEnd)}
+            </span>
+            {item.slice(substringEnd, item.length)}
+          </>
+        );
+      } else if (
+        highlightUserEntry &&
+        substring.length &&
+        substring.length <= item.length &&
+        filter === 'fuzzy'
+      ) {
+        // This will create a separate span for each character in the substring.
+        // This isn't ideal for every match, but it enables highlighting for fuzzy
+        // matches.
+        const lowerCaseSubstring = substring.toLocaleLowerCase();
+        return (
+          <>
+            {item.split('').map((char) => {
+              if (lowerCaseSubstring.includes(char.toLocaleLowerCase())) {
+                return <span class="autocomplete-partial">{char}</span>;
+              }
+              return char;
+            })}
+          </>
+        );
+      }
+      return item;
+    },
+    [highlightUserEntry, substring, filter]
+  );
+
+  const getRenderedItem = useCallback(
+    (item: Item, index: number) => {
+      let component;
+      if (typeof item === 'object') {
+        if (!itemTemplate) {
+          onError(`data must provide template for non-string items.`);
+          return null;
+        }
+        component = itemTemplate(item);
+      } else {
+        component = (
+          <div data-value={item}>{getItemChildren(item as string)}</div>
+        );
+      }
+      if (!isValidElement<ItemTemplateProps>(component)) {
+        return component;
+      }
+      const isDisabled = component.props['data-disabled'];
+      if (!component.props['data-value'] && !isDisabled) {
+        onError(
+          `expected a "data-value" or "data-disabled" attribute on the rendered template item.`
+        );
+      }
+      return cloneElement(component, {
+        'aria-disabled': isDisabled,
+        'aria-selected': activeIndex === index,
+        class: objStr({
+          'autocomplete-item': true,
+          [classes.autocompleteItem]: true,
+          [classes.autocompleteItemActive]: index === activeIndex,
+        }),
+        dir: 'auto',
+        id: getItemId(index),
+        key: item,
+        // unlike onClick, onMouseDown overrides the blur event handler
+        onMouseDown: handleMousedown,
+        part: 'option',
+        role: 'option',
+        ...component.props,
+      });
+    },
+    [
+      itemTemplate,
+      getItemId,
+      activeIndex,
+      classes,
+      handleMousedown,
+      getItemChildren,
+      onError,
+    ]
+  );
```

Comment:
I think it might be worthwhile to refactor this function in to a component in a separate file.